File: navic_time_processor.py
# navic_time_processor.py
# NAVIC (PX1125 GNSS) NMEA date/time parser — Linux & PyCharm friendly

from dataclasses import dataclass
import serial
import threading
from typing import Optional, List


@dataclass
class DateTime:
    year: Optional[int] = None
    month: Optional[int] = None
    day: Optional[int] = None
    hour: Optional[int] = None
    minute: Optional[int] = None
    second: Optional[int] = None

    def is_valid(self):
        return None not in (
            self.year, self.month, self.day,
            self.hour, self.minute, self.second
        )

    def __str__(self):
        if not self.is_valid():
            return "<Invalid DateTime>"
        return f"{self.year:04d}-{self.month:02d}-{self.day:02d} " \
               f"{self.hour:02d}:{self.minute:02d}:{self.second:02d}"


class UARTReader:
    """Simple pyserial wrapper."""
    def __init__(self, port: str, baud: int = 115200, timeout: float = 1.0):
        self.port = port
        self.baud = baud
        self.timeout = timeout
        self.ser = None

    def open(self):
        self.ser = serial.Serial(self.port, self.baud, timeout=self.timeout)

    def close(self):
        if self.ser:
            self.ser.close()

    def readline(self) -> Optional[str]:
        if not self.ser:
            return None

        raw = self.ser.readline()
        if not raw:
            return None

        return raw.decode(errors="ignore").strip()


class NavicTimeProcessor:
    """Parses NMEA strings into usable date/time objects."""

    def __init__(self, uart_port="/dev/ttyS0", baud=115200):
        self.uart = UARTReader(uart_port, baud)
        self.running = False
        self._thread = None

    @staticmethod
    def split(line: str) -> List[str]:
        return line.split(",")

    def parse_rmc(self, tokens: List[str]) -> Optional[DateTime]:
        # RMC: $GNRMC,hhmmss.sss,A,lat,N,lon,E,spd,track,ddmmyy,...
        if len(tokens) < 10:
            return None

        t = tokens[1]
        d = tokens[9]

        if len(t) < 6 or len(d) < 6:
            return None

        hour = int(t[0:2])
        minute = int(t[2:4])
        second = int(t[4:6])

        day = int(d[0:2])
        month = int(d[2:4])
        year = int(d[4:6]) + 2000

        return DateTime(year, month, day, hour, minute, second)

    def parse_gga(self, tokens: List[str]) -> Optional[DateTime]:
        # GGA contains only time.
        if len(tokens) < 2:
            return None

        t = tokens[1]
        if len(t) < 6:
            return None

        hour = int(t[0:2])
        minute = int(t[2:4])
        second = int(t[4:6])

        # date unknown
        return DateTime(None, None, None, hour, minute, second)

    def process_line(self, line: str) -> Optional[DateTime]:
        if not line.startswith("$"):
            return None

        payload = line[1:].split("*")[0]
        tokens = self.split(payload)

        header = tokens[0]

        if header.endswith("RMC"):
            return self.parse_rmc(tokens)

        if header.endswith("GGA"):
            return self.parse_gga(tokens)

        return None

    def start(self, callback):
        """Starts background reading of UART + callback(DateTime)."""
        self.uart.open()
        self.running = True

        def worker():
            while self.running:
                line = self.uart.readline()
                if line:
                    dt = self.process_line(line)
                    if dt and dt.is_valid():
                        callback(dt)

        self._thread = threading.Thread(target=worker, daemon=True)
        self._thread.start()

    def stop(self):
        self.running = False
        if self._thread:
            self._thread.join(timeout=1)
        self.uart.close()

File: navic_time_processor.cpp
#include <include/navic_subsystem/navic_time_processor.h>


namespace navic
{
    navic_time_processor::navic_time_processor()
    {
        this->token_vector.clear();

        this->uart_channel_handle = std::make_unique<uart_channel>("/dev/ttyS0",
                                                                   B115200,
                                                                   8,
                                                                   Parity::None,
                                                                   1
                                                                  );
    }

    void  navic_time_processor::clear_date_time()
    {
        this->parsed_date_time.hour = 0;
        this->parsed_date_time.minute = 0;
        this->parsed_date_time.second = 0;

        this->parsed_date_time.day = 0;
        this->parsed_date_time.month = 0;
        this->parsed_date_time.year = 0;
    }

    // Parse an NMEA sentence and extract time/date if available
    DateTime& navic_time_processor::parse(const std::string& sentence)
    {
        if (sentence.empty() || sentence[0] != '$')
        {
            throw std::invalid_argument("Invalid NMEA sentence");
        }

        split(sentence, ',');

        if (token_vector[0].find("RMC") != std::string::npos)
        {
            parseRmc();
        }
        else if (token_vector[0].find("ZDA") != std::string::npos)
        {
            parseZda();
        }
        else
        {
            throw std::invalid_argument("Unsupported NMEA sentence type");
        }

        return this->parsed_date_time;
    }

     // Split helper
    void navic_time_processor::split(const std::string& str, char delimiter)
    {
        std::stringstream ss(str);
        std::string item;

        this->token_vector.clear();

        while (std::getline(ss, item, delimiter))
        {
            this->token_vector.push_back(item);
        }
    }

    // Parse RMC sentence: $GNRMC,hhmmss.sss,A,...,ddmmyy,...
    void navic_time_processor::parseRmc()
    {
        this->clear_date_time();

        //less than 10 fields, then throw exception error.
        if (this->token_vector.size() < 10)
        {
            throw std::invalid_argument("Incomplete RMC sentence");
        }

        std::string timeStr = this->token_vector[1]; // hhmmss.sss
        std::string dateStr = this->token_vector[9]; // ddmmyy

        this->parsed_date_time.hour   = std::stoi(timeStr.substr(0, 2));
        this->parsed_date_time.minute = std::stoi(timeStr.substr(2, 2));
        this->parsed_date_time.second = std::stoi(timeStr.substr(4, 2));

        this->parsed_date_time.day    = std::stoi(dateStr.substr(0, 2));
        this->parsed_date_time.month  = std::stoi(dateStr.substr(2, 2));
        this->parsed_date_time.year   = 2000 + std::stoi(dateStr.substr(4, 2)); // yy → 20yy
    }

    // Parse ZDA sentence: $GNZDA,hhmmss.sss,dd,mm,yyyy,...
    void navic_time_processor::parseZda()
    {
        this->clear_date_time();

        if (this->token_vector.size() < 5)
        {
            throw std::invalid_argument("Incomplete ZDA sentence");
        }

        std::string timeStr = this->token_vector[1]; // hhmmss.sss

        this->parsed_date_time.hour   = std::stoi(timeStr.substr(0, 2));
        this->parsed_date_time.minute = std::stoi(timeStr.substr(2, 2));
        this->parsed_date_time.second = std::stoi(timeStr.substr(4, 2));

        this->parsed_date_time.day    = std::stoi(this->token_vector[2]);
        this->parsed_date_time.month  = std::stoi(this->token_vector[3]);
        this->parsed_date_time.year   = std::stoi(this->token_vector[4]);
    }

}
